### ISSUE TITLE
False positive when a field cannot be None

### ISSUE THREAD

**Issue by edchu** [microsoft/pyright#8428]
When checking whether a variable is None with another function, it seems pyright still thinks the variable can be None

Consider this example:
```python
def is_none(the_str: str | None) -> bool:
    if str is None:
        return True

    return False


def test_f(some_string: str | None) -> str:
    if is_none(some_string) is True:
        return ""

    return some_string.lower()
```

pyright complains with the following:
- error: "lower" is not a known attribute of "None" (reportOptionalMemberAccess)

This is with pyright version 1.1.371

**Comment by erictraut** [comment]
Pyright's behavior is correct here. Types are narrowed only within a local scope. So the fact that you're narrowing the type of `the_str` within `is_none`, this doesn't affected the narrowed type of `some_string` in `test_f`.

If you want `is_none` to act as a user-defined type guard function, you need to change its return type to `TypeIs[None]`. For more details, refer to [this section of the typing spec](https://typing.readthedocs.io/en/latest/spec/narrowing.html#typeis).

Here's how you would need to change your code for this to type check without errors:

```python
from typing_extensions import TypeIs

def is_none(the_str: str | None) -> TypeIs[None]:
    return the_str is None

def test_f(some_string: str | None) -> str:
    if is_none(some_string):
        return ""

    return some_string.lower()
```
